### ISSUE TITLE
Role is executed multiple times when referenced indirectly as dependency

### ISSUE THREAD

**Issue by zarnovican** [ansible/ansible#14046]
##### Issue Type:

Bug Report
##### Ansible Version:

```
ansible 2.0.0.2
  config file = /home/zarnovic/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

No changes.
##### Environment:

Independent.

for completeness:
Control host: Fedora 23
Managed host: Debian Jessie
##### Summary:

Role is executed multiple times when referenced indirectly as dependency.
##### Steps To Reproduce:

Create three roles "a", "b", "c" with following dependencies:
"b" depends on "a"
"c" depends on "a", "b"

``` bash
mkdir -p roles/{a,b,c}/{tasks,meta}
for r in a b c; do printf -- '---\n- debug: msg="task '$r'"\n' > roles/$r/tasks/main.yaml; done
printf -- '---\ndependencies:\n  - a\n' > roles/b/meta/main.yaml
printf -- '---\ndependencies:\n  - a\n  - b\n' > roles/c/meta/main.yaml
```

Include only role "c" in playbook and execute:

```

---
- name: 'playbook'
  hosts: all
  gather_facts: no

  roles:
    - c
```
##### Expected Results:

Each role should be included only once. Three tasks will be executed.

```
...
PLAY RECAP ******************************************************************** 
vm1                        : ok=3    changed=0    unreachable=0    failed=0   
```

This is the behavior of Ansible 1.9.4
##### Actual Results:

Role "a" is included twice.

```
$ ansible-playbook -i inventory.ini site.yaml -u zarnovic

PLAY [playbook] ****************************************************************

TASK [a : debug] ***************************************************************
ok: [vm1] => {
    "msg": "task a"
}

TASK [a : debug] ***************************************************************
ok: [vm1] => {
    "msg": "task a"
}

TASK [b : debug] ***************************************************************
ok: [vm1] => {
    "msg": "task b"
}

TASK [c : debug] ***************************************************************
ok: [vm1] => {
    "msg": "task c"
}

PLAY RECAP *********************************************************************
vm1                        : ok=4    changed=0    unreachable=0    failed=0   
```

Note:
- when playbook includes "a", "c", role "a" is included three times
- when playbook includes "b", "c", role "a" is included only once

**Comment by jimi-c** [comment]
This is an interesting bug, as it's hitting a corner case. 

The task iterator class checks to see if the currently iterated task has a role, and if so then checks to see if the current task state had a role. If the new tasks role != the current state's role, it checks to see if that role had any tasks run from it and if so marks the role as complete. The problem here is the progression of roles -> role A, role A, role B, role C. 

So in this case, despite the fact that the role came from a different inheritance path, as it moves on it's the same role and it doesn't mark the role as complete. After it runs the second time, the new role is role B, and it's marked complete. This shouldn't be too difficult to fix, working on it now.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by zarnovican** [comment]
@jimi-c: Thanks for the quick turnaround!

I can confirm that the problem is fixed for the small reproducer I provided.

However..

This change have broken something terribly (git bisect-ed it). On my real playbook, some tasks are now skipped entirely :-1: ! (53 tasks vs 64). By the quick look, it seems the problem is with `include`

```
# role/foo/tasks/main.yaml
- name: task before

- include: bar.yaml

- name: task after
```

Looks like, included tasks and all the following tasks are skipped.

**Comment by jimi-c** [comment]
@zarnovican where is the include in the dependency chain?

**Comment by jimi-c** [comment]
@zarnovican pushed a8ffa021344b5548813790f18dda68d32717a4d7 to fix this. Also validated multiple levels of nested includes work as expected.

Thanks for catching this!

**Comment by zarnovican** [comment]
@jimi-c Thanks for a quick fix. Looks good to me, now.

Just a heads-up, there is something fishy in the task iterator, performance-wise. I'm still getting sluggish performance (50s vs 14s) on Ansible 2.0/devel with role dependencies. My initial assumption, that because of this bug, tasks are executed multiple times, turned out to be wrong (it's still slow). I will open a separate ticket when I'll have a nice reproducer.

**Comment by jimi-c** [comment]
@zarnovican running with ANSIBLE_DEBUG=1 will show some timing info, might help show where the slowness is. Right now almost all of the debug statements for that are in the executor portions of the code though, so if it's inventory or somewhere else it may not be helpful. We also don't have debug verbosity levels, so it's a LOT of info dumped out.
